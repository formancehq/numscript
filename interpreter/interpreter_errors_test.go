package interpreter_test

import (
	"testing"

	"github.com/formancehq/numscript/interpreter"
	"github.com/formancehq/numscript/parser"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
)

func matchErrWithSnapshots(t *testing.T, src string) {
	parsed := parser.Parse(src)

	_, err := interpreter.RunProgram(parsed.Value, interpreter.RunProgramOptions{})
	require.NotNil(t, err)
	snaps.MatchSnapshot(t, err.GetRange().ShowOnSource(parsed.Source))
}

func TestShowUnboundVar(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = $unbound_var
  destination = @dest
)`)

}

func TestShowMissingFundsSingleAccount(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = @a
  destination = @dest
)`)

}

func TestShowMissingFundsInorder(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = {
    @a
    @b
	}
  destination = @dest
)`)
}

func TestShowMissingFundsAllotment(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = {
    1/2 from @a
		remaining from @world
	}
  destination = @dest
)`)
}

func TestShowMissingFundsMax(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = max [COIN 2] from {
    1/2 from @world
    remaining from @world
  }
  destination = @dest
)`)
}
