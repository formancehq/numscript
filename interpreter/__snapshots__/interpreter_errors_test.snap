
[TestShowUnboundVar - 1]
  1 |   source = $unbound_var
    |            ~~~~~~~~~~~~
---

[TestShowMissingFundsSingleAccount - 1]
  1 |   source = @a
    |            ~~
---

[TestShowMissingFundsInorder - 1]
  1 |   source = {
    |            ~
  2 |     @a
    | ~~~~~~
  3 |     @b
    | ~~~~~~
  4 |  }
    | ~~
---

[TestShowMissingFundsAllotment - 1]
  2 |     1/2 from @a
    |              ~~
---

[TestShowMissingFundsMax - 1]
  1 |   source = max [COIN 2] from {
    |            ~~~~~~~~~~~~~~~~~~~
  2 |     1/2 from @world
    | ~~~~~~~~~~~~~~~~~~~
  3 |     remaining from @world
    | ~~~~~~~~~~~~~~~~~~~~~~~~~
  4 |   }
    | ~~~
---
